tests/integration/security/reachability/reachability_test.go

```diff
@@ -44,280 +44,225 @@ import (
-							ExpectSuccess: true,
-						},
-						{
-							From: a,
-							Options: echo.CallOptions{
-								Target:   headless,
-								PortName: "http",
-								Scheme:   scheme.HTTP,
-							},
-							ExpectSuccess: true,
-						},
-						{
-							From: a,
-							Options: echo.CallOptions{
-								Target:   headless,
-								PortName: "grpc",
-								Scheme:   scheme.GRPC,
-							},
-							ExpectSuccess: true,
-						},
-						{
-							From: naked,
-							Options: echo.CallOptions{
-								Target:   b,
-								PortName: "http",
-								Scheme:   scheme.HTTP,
-							},
-							ExpectSuccess: false,
-						},
-						{
-							From: naked,
-							Options: echo.CallOptions{
-								Target:   b,
-								PortName: "grpc",
-								Scheme:   scheme.GRPC,
-							},
-							ExpectSuccess: false,
-						},
-						{
-							From: naked,
-							Options: echo.CallOptions{
-								Target:   b,
-								PortName: "tcp",
-								Scheme:   scheme.HTTP,
-							},
-							ExpectSuccess: false,
-						},
+					configFile:          "global-mtls-on.yaml",
+					namespace:           systemNS,
+					requiredEnvironment: environment.Kube,
+					include: func(src echo.Instance, opts echo.CallOptions) bool {
+						// Exclude calls to the headless TCP port.
+						if opts.Target == headless && opts.PortName == "tcp" {
+							return false
+						}
+
+						// Exclude headless->headless
+						return src != headless || opts.Target != headless
+					},
+					expectSuccess: func(src echo.Instance, opts echo.CallOptions) bool {
+						if src == naked && opts.Target == naked {
+							// naked->naked should always succeed.
+							return true
+						}
+
+						// If one of the two endpoints is naked, expect failure.
+						return src != naked && opts.Target != naked
 					},
 				},
 				{
-					configFile: "global-mtls-permissive.yaml",
-					namespace:  systemNS,
-					subTests: []connection.Checker{
-						{
-							From: a,
-							Options: echo.CallOptions{
-								Target:   b,
-								PortName: "http",
-								Scheme:   scheme.HTTP,
-							},
-							ExpectSuccess: true,
-						},
-						{
-							From: naked,
-							Options: echo.CallOptions{
-								Target:   b,
-								PortName: "http",
-								Scheme:   scheme.HTTP,
-							},
-							ExpectSuccess: true,
-						},
-						{
-							From: naked,
-							Options: echo.CallOptions{
-								Target:   b,
-								PortName: "grpc",
-								Scheme:   scheme.GRPC,
-							},
-							ExpectSuccess: true,
-						},
-						{
-							From: naked,
-							Options: echo.CallOptions{
-								Target:   b,
-								PortName: "tcp",
-								Scheme:   scheme.HTTP,
-							},
-							ExpectSuccess: true,
-						},
+					configFile:          "global-mtls-permissive.yaml",
+					namespace:           systemNS,
+					requiredEnvironment: environment.Kube,
+					include: func(src echo.Instance, opts echo.CallOptions) bool {
+						// Exclude calls to the naked app.
+						if opts.Target == naked {
```

Comment:
optional. We may remove this (include naked target) to make it simple. We didn't have that test case before as it's kinda tedious to copy-paste, but with this change it's about the same.

Reply:
it actually doesn't work today.  The echo app only supports plain HTTP, so without a sidecar there is no negotiation